files/routes/votes.py

```diff
@@ -56,62 +56,72 @@ def admin_vote_info_get(v):
+	if post.author_id in {AUTOPOLLER_ID,AUTOBETTER_ID,AUTOCHOICE_ID}: return {"error": "forbidden."}, 403
+	
+	# get the old vote, if we have one
+	vote = g.db.query(Vote).filter_by(user_id=v.id, submission_id=post.id).one_or_none()
 
-	coin_delta = 1
-	if v.id == post.author.id:
-		coin_delta = 0
+	# should we just do nothing? we could just do nothing
+	if vote and vote.vote_type == new: return "", 204
 
-	if existing and existing.vote_type == new: return "", 204
+	# at this point we are guaranteed to be making a change
 
-	if existing:
-		if existing.vote_type == 0 and new != 0:
-			post.author.coins += coin_delta
-			post.author.truecoins += coin_delta
-			g.db.add(post.author)
-			existing.vote_type = new
-			g.db.add(existing)
-		elif existing.vote_type != 0 and new == 0:
-			post.author.coins -= coin_delta
-			post.author.truecoins -= coin_delta
-			g.db.add(post.author)
-			g.db.delete(existing)
-		else:
-			existing.vote_type = new
-			g.db.add(existing)
-	elif new != 0:
-		post.author.coins += coin_delta
-		post.author.truecoins += coin_delta
-		g.db.add(post.author)
+	# author votes don't matter if it's the author themselves
+	points_matter = (v.id != post.author.id)
 
+	if vote:
+		# remove the old score data
+		if points_matter:
+			post.author.coins -= vote.vote_type
+			post.author.truecoins -= vote.vote_type
+			# we'll be saving later anyway, so don't bother doing so here
+	else:
+		# create new vote data
 		DEFAULT_IMAGE = '/assets/images/default-profile-pic.webp'
 
-		if new == 1 and (v.shadowbanned or (v.is_banned and not v.unban_utc) or (v.profile_url == DEFAULT_IMAGE and not v.customtitle and v.namecolor == DEFAULT_COLOR)): real = False
-		else: real = True
-
-		vote = Vote(user_id=v.id,
-					vote_type=new,
-					submission_id=post_id,
-					app_id=v.client.application.id if v.client else None,
-					real = real
+		vote = Vote(
+						user_id=v.id,
+						vote_type=new,
+						submission_id=post_id,
+						app_id=v.client.application.id if v.client else None,
 					)
-		g.db.add(vote)
-
+	
+	# update the vote data
+	vote.vote_type = new
+
+	real = True
+	if v.shadowbanned: real = False
+	if v.is_banned and not v.unban_utc: real = False
+	vote.real = real
+
+	# add relevant points
+	if points_matter:
+		post.author.coins += vote.vote_type
+		post.author.truecoins += vote.vote_type
+	
+	# database it up
+	g.db.add(post.author)
+	g.db.add(vote)
+	
+	# update post stats (this is horrendously slow?!)
 	g.db.flush()
 	post.upvotes = g.db.query(Vote.submission_id).filter_by(submission_id=post.id, vote_type=1).count()
 	post.downvotes = g.db.query(Vote.submission_id).filter_by(submission_id=post.id, vote_type=-1).count()
-	post.realupvotes = g.db.query(Vote.submission_id).filter_by(submission_id=post.id, real=True).count()
-	if post.author.progressivestack: post.realupvotes *= 2
+	post.realupvotes = g.db.query(Vote.submission_id).filter_by(submission_id=post.id, vote_type=1, real=True).count()
```

Comment:
Based on the way we use `realupvotes` to sort [here](https://github.com/themotte/rDrama/blob/frost/files/routes/front.py#L347) (and similarly for comments in routes/posts.py), I think this calculation needs to take downvotes into account too.

Reply:
So I actually think that `realupvotes` was never meant to include downvotes; it actually is "real upvotes", the name is accurate. Note that "top", "bottom", "controversial", and so forth all take downvotes into account separately, and I believe a similar thing is happening here. I haven't sat down to figure out whatever the hell is going on with that "top" equation, but I think it is literally *just* meant to consider upvotes, not downvotes.

(There's a good chance these are all getting changed at some point :V)